src/Core/Invoker.php

```diff
@@ -0,0 +1,152 @@
+<?php
+declare(strict_types=1);
+
+/**
+ * CakePHP(tm) : Rapid Development Framework (https://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ * @link          https://cakephp.org CakePHP(tm) Project
+ * @since         5.0.0
+ * @license       https://opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\Core;
+
+use Closure;
+use InvalidArgumentException;
+use ReflectionFunction;
+use ReflectionNamedType;
+
+class Invoker
+{
+    /**
+     * @var \Cake\Core\ContainerInterface
+     */
+    protected ContainerInterface $container;
+
+    /**
+     * @param \Cake\Core\ContainerInterface $container Containter instance
+     */
+    public function __construct(ContainerInterface $container)
+    {
+        $this->container = $container;
+    }
+
+    /**
+     * Invokes callable with supplied arguments and loads missing
+     * arguments from the container if the type is valid.
+     *
+     * @param \Closure $closure Function to invoke
+     * @param array<mixed> $arguments Function arguments
+     * @return mixed
+     * @throws \InvalidArgumentException When unable to find argument for parameters.
+     */
+    public function invoke(Closure $closure, array $arguments): mixed
+    {
+        return $closure(...$this->resolveArguments($closure, $arguments));
+    }
+
+    /**
+     * Invokes callable with supplied arguments and loads missing
+     * arguments from the container if the type is valid.
+     *
+     * String arguments are coerced to parameter type for
+     * float, int and bool types.
+     *
+     * @param \Closure $closure Function to invoke
+     * @param array<mixed> $arguments Function arguments
+     * @return mixed
+     * @throws \InvalidArgumentException When unable to find argument for parameters.
+     */
+    public function invokeWithCoercion(Closure $closure, array $arguments)
```

Comment:
When would you not want coercion? Seems like we can have a simpler interface if we always coerce.

Reply:
I assumed there was a use case for a generic invoker. Maybe we don't need an Invoker class and just a general way to resolve arguments.

Reply:
I don't think we have that use case yet. We could add the methods when we understand those requirements better. As you know, public APIs are hard for us to evolve. You could mark this class as `internal` to escape our backwards compat promises.